app/src/main/java/com/davidtomas/taskyapp/features/auth/presentation/login/LoginViewModel.kt

```diff
@@ -1,5 +1,87 @@
+import com.davidtomas.taskyapp.features.auth.domain.model.EmailFormatError
+import com.davidtomas.taskyapp.features.auth.domain.model.EmailLoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.model.LoginInputValidationErrors
+import com.davidtomas.taskyapp.features.auth.domain.model.LoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.model.MissingMandatoryField
+import com.davidtomas.taskyapp.features.auth.domain.model.PasswordFormatError
+import com.davidtomas.taskyapp.features.auth.domain.model.PasswordLoginValidationType
+import com.davidtomas.taskyapp.features.auth.domain.useCase.ValidateLoginInputsUseCase
+
+class LoginViewModel(
+    private val validateLoginInputsUseCase: ValidateLoginInputsUseCase
+) : ViewModel() {
+
+    var state by mutableStateOf(LoginState())
+        private set
+
+    fun onAction(action: LoginAction) {
+        when (action) {
+            is OnChangePasswordVisibility -> {
+                state = state.copy(isPasswordVisible = !state.isPasswordVisible)
+            }
+
+            is OnEmailChanged -> {
+                state = state.copy(email = action.email)
+                validateInputs(EmailLoginValidationType(action.email))
+            }
+
+            is OnPasswordChanged -> {
+                state = state.copy(password = action.password)
+                validateInputs(PasswordLoginValidationType(action.password))
+            }
+
+            else -> Unit
+        }
+    }
+
+    private fun validateInputs(
+        loginValidationType: LoginValidationType
+    ) {
+        validateLoginInputsUseCase.invoke(loginValidationType).fold(
+            fnL = ::onValidationError,
+            fnR = {
+                onValidationSuccess(loginValidationType)
+            }
+        )
+    }
+
+    private fun onValidationSuccess(loginValidationType: LoginValidationType) {
+        when (loginValidationType) {
+            is AllFieldsLoginValidationType -> state = state.copy(isLoginBtnEnabled = true)
+            else -> {
+                state = state.copy(
+                    emailErrMsg = String.EMPTY_STRING,
+                    passwordErrMsg = String.EMPTY_STRING
+                )
+                validateInputs(AllFieldsLoginValidationType(state.email, state.password))
+            }
+        }
+    }
+
+    private fun onValidationError(failure: Failure) {
+        state = state.copy(isLoginBtnEnabled = false)
+        val validationsErrors = (failure as LoginInputValidationErrors).validationErrors
+        validationsErrors.forEach { loginInputValidationError ->
+            when (loginInputValidationError) {
+                EmailFormatError -> state =
+                    state.copy(emailErrMsg = "The format of the email is wrong")
+
+                PasswordFormatError -> state =
+                    state.copy(passwordErrMsg = "The format of the password is wrong")
```

Comment:
Aren't you working with string resources? Then why hardcode strings here without being able to localize them?

Reply:
I have some doubts about this. On thursday you talked about the UIText util, but I havent understand it at all. It is necessary tu use it here in this viewModel. The problem is that is I make my state fields of type UIText when I do the validations I need to send the actual data in the fields... I am a bit confused about this. Heelp  😭  hahaha

Reply:
`UiText` is usually only needed for static strings you create in the ViewModel, e.g. when interpreting an error code and parsing it to a string resource. It doesn't really have anything to do with sending data to the validation use cases, as you have these as raw strings (these are user-made, so you don't need any string resources).
UI -> ViewModel: Normal strings
ViewModel -> UI: UiText

Reply:
Very well explained, i now understood, thank you very much